### PR TITLE
Added automatic serialization for simple structs (see #3034)

### PR DESCRIPTION
**Pull request by wteiwewte**
- Added automatic serialization for simple, brace-initializable structs. Every struct's field must be serializable and public.
  - Added one unit test for automatic serialization and removed fail_compile_non_serializable_class test (now it compiles, so no need to have it)
-  Arity function doesn't support copyable/movable structs with one field (because then exists constructor with one argument)
- Up to 15 fields (can be extended with simple copy&paste)

### CODE REVIEW

File: hpx/runtime/serialization/access.hpp [hunk 1]
```diff
@@ -76,15 +108,34 @@ namespace hpx { namespace serialization
...
-                // finds only serialize-member function and doesn't
-                // perform ADL
+                template<class T1> struct dependent_false : std::false_type {};
+
+
                 template <class Archive>
                 static void call(Archive& ar, T& t, unsigned)
                 {
-                    detail::serialize_force_adl(ar, t, 0);
+                    if constexpr (has_serialize_adl<T>::value)
```
Comment (hkaiser) [inline hpx/runtime/serialization/access.hpp]: Could you please guard the use of `if constexpr` with a feature test macro. The whole new serialization functionality should be made available when compiling in C++17 mode only anyways, so enabling `if constexpr` at the same time should be ok.

**Comment by wteiwewte** [comment]
- Added `__cpp_if_constexpr` test macro.

- `arity` function checks if there is no public constructor with given number of arguments. `hpx::util::tuple` has variadic constructor, so for every number of arguments it has public constructor, thus with `hpx::util::tuple` it won't work (same for `std::tuple`)

File: hpx/runtime/serialization/access.hpp (hunk 1 above)
Comment (hkaiser) [reply hpx/runtime/serialization/access.hpp]: `__cpp_if_constexpr` is not supported when compiling with a C++11 compiler. You might want to add a feature test here: https://github.com/STEllAR-GROUP/hpx/tree/master/cmake/tests, https://github.com/STEllAR-GROUP/hpx/blob/master/cmake/HPX_PerformCxxFeatureTests.cmake#L198, and here: https://github.com/STEllAR-GROUP/hpx/blob/master/cmake/HPX_AddConfigTest.cmake#L661. You can then use the preprocessor macro to hide the new code from older compilers.

**Comment by hkaiser** [comment]
@AntonBikineev is this something you'd be interested in reviewing?

**Comment by hkaiser** [comment]
@wteiwewte could you please add the missing license/copyright as reported here: https://55221-4455628-gh.circle-artifacts.com/0/hpx_inspect_report.html? The other reported errors seem to be unrelated.

**Comment by wteiwewte** [comment]
Sorry for the delay. I added missing licenses and feature tests (through cmake).

**Comment by hkaiser** [comment]
@wteiwewte inspect is still unhappy with your proposed changes (see: https://circleci.com/gh/STEllAR-GROUP/hpx/57135). Additionally, could you please rebase this branch to resolve the conflicts with master?

**Comment by hkaiser** [comment]
@wteiwewte Thanks a lot for working on this! @msimberg I think this is good to go now.

**Comment by hkaiser** [comment]
@wteiwewte Congrats to your first accepted PR! Please get in contact with @aserio if you would like to receive a STE||AR-Group t-shirt. We give that away to our first-time contributors.

**Pull request merged**